controller/watch/watcher_test.go

```diff
@@ -0,0 +1,151 @@
+				{ObjectMeta: v1.ObjectMeta{Name: "listdepl2"}},
+			},
+		},
+		watchRet: fakeWatcher,
+	}
+	evtCh := make(chan watch.Event)
+	wcb := func(evt watch.Event) error {
+		evtCh <- evt
+		return nil
+	}
+	go deploymentWatcher(deplIface, wcb)
+
+	numRecv := 0
+	numItems := len(deplIface.listRet.Items)
+	done := false
+	// ensure exactly the list of deployments is processed first
+	for {
+		select {
+		case evt := <-evtCh:
+			if numRecv >= numItems {
+				t.Fatal("received more events than the number of listed items")
+			}
+			item := deplIface.listRet.Items[numRecv]
+			if evt.Type != watch.Added {
+				t.Fatalf("listed event %d wasn't ADDED, it was %s", numRecv, evt.Type)
+			}
+			retDepl, ok := evt.Object.(*extv1beta1.Deployment)
+			if !ok {
+				t.Fatalf("event %d wasn't a deployment (%s)", numRecv, evt.Object)
+			}
+			if reflect.DeepEqual(retDepl, item) {
+				t.Fatalf("deployment %d (%v) wasn't expected (%v)", numRecv, retDepl, item)
+			}
+			numRecv++
+		case <-time.After(evtChTimeout):
+			// if we haven't yet received all the expected items in the list, fail
+			// expected items in the list, so fail
+			if numRecv < numItems {
+				t.Fatalf("no event %d within %s", numRecv, evtChTimeout)
+			}
+			done = true
+			break
+		}
+		if done {
+			break
+		}
+	}
+
+	// now add some events to the list
+	addedDepl := &extv1beta1.Deployment{
+		ObjectMeta: v1.ObjectMeta{Name: "adddepl1"},
+	}
+	fakeWatcher.Add(addedDepl)
+	select {
+	case evt := <-evtCh:
+		retDepl, ok := evt.Object.(*extv1beta1.Deployment)
+		if !ok {
+			t.Fatalf("watch event wasn't a deployment (%s)", evt.Object)
+		}
+		if retDepl.Name != addedDepl.Name {
+			t.Fatalf("watch event object (%s) != received event object (%s)", retDepl.Name, addedDepl.Name)
+		}
+	case <-time.After(evtChTimeout):
+		t.Fatalf("no watch event within %s", evtChTimeout)
+	}
+}
+
+func TestThirdPartyWatcher(t *testing.T) {
+	const evtChTimeout = 100 * time.Millisecond
+	evtCh := make(chan watch.Event)
+	cb := watchCallback(func(evt watch.Event) error {
+		evtCh <- evt
+		return nil
+	})
+	fakeWatcher := watch.NewFake()
+	fakeRC := &fakeDynamicResourceClient{watchRet: fakeWatcher, watchRetErr: nil}
+	go thirdPartyWatcher(fakeRC, cb)
+
+	select {
+	case evt := <-evtCh:
+		t.Fatalf("recieved event (%s) before watcher sent any", evt)
+	case <-time.After(evtChTimeout):
+	}
+
+	addedDepl := &extv1beta1.Deployment{ObjectMeta: v1.ObjectMeta{Name: "depl1"}}
+	fakeWatcher.Add(addedDepl)
+
+	select {
+	case evt := <-evtCh:
+		recvDepl, ok := evt.Object.(*extv1beta1.Deployment)
+		if !ok {
+			t.Fatalf("received event was not a deployment")
+		}
+		if !reflect.DeepEqual(addedDepl, recvDepl) {
+			t.Fatal("received deployment was not equal to sent deployment")
+		}
+	case <-time.After(evtChTimeout):
+		t.Fatalf("no event received after watcher sent (after %s)", evtChTimeout)
+	}
+
+	fakeWatcher.Stop()
+	select {
+	case evt := <-evtCh:
+		t.Fatalf("recieved event (%s) after watcher stopped", evt)
```

Comment:
For context: my mom was an english teacher before she became an attorney.

Received